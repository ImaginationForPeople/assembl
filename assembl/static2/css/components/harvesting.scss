.harvesting-anchor {
  position: absolute;
  z-index: 1000;
}

.harvesting-container {
  position: absolute;
  left: 0;
  z-index: 99;
}

.left-side-harvesting-button {
  &:hover {
    .assembl-icon-catch:before {
      font-weight: bold;
    }

    .left-side-harvesting-button__button__inside {
      background-color: #ffed5b;
    }
  }

  .left-side-harvesting-button__button {
    background: white;
    border: 1px solid #ccc;
    border-radius: 50%;
    width: 44px;
    height: 44px;
    box-shadow: 2px 2px 4px 0 #ccc;
    cursor: pointer;

    .left-side-harvesting-button__button__inside {
      text-align: center;
      margin: 4px;
      width: calc(100% - 8px);
      height: calc(100% - 8px);
      background-color: #fff7b5;
      border-radius: 50%;
      display: flex;
    }
  }
}

.confirm-harvest-button {
  text-align: center;
  padding: 1px 0 0 6px;
  font-size: 20px;

  &:before {
    margin-right: 2px;
    font-size: $font-size-sm;
  }
}

.harvesting-mode-on {
  .extracts {
    display: none;
  }

  .post-body--is-harvestable {
    /* FIXME: Factorizing these ::selection rules cancels effect on Firefox */
    .post-body-content *::selection {
      background: $harvesting-yellow;
    }

    .post-body-content *::-moz-selection {
      background: $harvesting-yellow;
    }
  }
}

.active-box {
  border-left-color: $harvesting-green;
}

.box-icon {
  width: 34px;
  height: 34px;
  border: 3px solid $harvesting-green;
  border-radius: 50%;
  position: absolute;
  background-color: $white;
  margin-top: 15px;
  margin-left: 3px;

  svg {
    margin: 4px 0 0 4px;
  }
}

.box-icon-2 {
  margin-top: 40px;
  z-index: 2;
}

.harvesting-box {
  box-shadow: 2px 2px 4px 0 $grey1;
  margin: 0 0 20px 18px;
  width: 295px;

  .harvesting-box-header {
    padding: 16px 20px 15px 20px;
    border-bottom: 1px solid $grey1;
  }

  .harvesting-status {
    display: inline-block;
    font-size: $font-size-xs;

    .harvesting-in-progress {
      color: $black;
      margin-left: -12px;
      padding: 0 0 15px;
    }

    .validated-harvesting {
      color: $harvesting-green;
      margin-left: -12px;
      padding: 0 0 15px;
    }

    .harvesting-status-label {
      display: inline-block;
      vertical-align: sub;
      margin-left: 15px;
      font-size: $font-size-xxs;
      margin-top: 2px;
      width: 145px;
      height: 30px;
      line-height: 30px;
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
    }

    .harvesting-taxonomy-label {
      display: inline-block;
      vertical-align: sub;
      margin-left: 15px;
      font-size: $font-size-xxs;
      margin-top: 2px;
      width: 145px;
      height: 30px;
    }
  }

  .profile {
    font-size: 11px;
    color: $grey3;
    margin-left: -5px;
    display: flex;

    img {
      border-radius: 50%;
      margin: 0 10px 0 6px;
    }

    .assembl-icon-profil {
      vertical-align: bottom;

      &:before {
        font-size: 30px;
        margin-right: 10px;
      }
    }

    .harvesting-date {
      font-family: $font-family4;
      font-size: $font-size-xxs;
    }

    .harvesting-infos {
      display: inline-block;
      vertical-align: bottom;

      .username {
        width: 208px;
        text-overflow: ellipsis;
        white-space: nowrap;
        overflow: hidden;
      }
    }
  }

  .button-bar {
    padding-top: 5px;
    float: right;
    margin-right: -25px;

    .taxonomy-menu-btn {
      border: none;
      padding: 0 0 0 2px;
      background: none;

      &:hover,
      &:focus {
        background: none !important;
      }

      .assembl-icon-ellipsis-vert {
        &:before {
          font-size: 20px;
          vertical-align: middle;
          padding: 2px 0 0;
        }

        &:hover {
          color: $harvesting-green;
        }
      }
    }

    .btn[disabled] {
      opacity: 1;
      border-color: $grey5;

      span {
        color: $grey5;
      }

      &:hover {
        background-color: $white;
        border-color: $grey5;

        span {
          color: $grey5;
        }
      }
    }

    .btn.active {
      background-color: $white;
      border-color: $second-color;
      box-shadow: none;

      span {
        color: $second-color;
      }
    }

    .btn-default {
      margin: 0 3px;
      border-radius: 5px;
      width: 26px;
      height: 26px;
      border-color: $grey3;

      span {
        color: $grey3;
      }

      &:hover {
        background-color: $white;
        border-color: $second-color;

        span {
          color: $second-color;
        }
      }

      &:focus {
        background-color: $white;
        outline: none;
      }

      &:active {
        box-shadow: none;
      }

      span {
        &:before {
          font-size: 12px;
          display: block;
          margin: 0 0 0 -6px;
        }
      }
    }

    .taxonomy-menu {
      font-size: $font-size-xs;
      letter-spacing: 1.3px;
      line-height: 1;
      border: 1px solid $grey1;
      border-radius: 0;
      position: absolute;
      left: 295px;
      top: 25px;

      .assembl-icon-cancel {
        cursor: pointer;
        position: absolute;
        right: 5px;
        top: 10px;
        font-size: 13px;
      }

      .button-taxonomy {
        width: 90%;
        border-radius: 0;
        height: auto;
        padding: 3px 0;
        margin: 10px 0;
      }

      .button-taxonomy:hover {
        border-color: $first-color;

        span {
          color: white;
        }
      }

      .arrow {
        display: none;
      }

      .popover-content {
        padding: 30px 0 10px 0;
        width: 245px;
      }

      .taxonomy-label {
        color: $black;
        padding: 5px 10px 10px 8px;
        font-family: $font-family4;

        .nature-label {
          display: contents;
          position: absolute;
          margin-top: 5px;
          padding-left: 10px;

          span {
            display: block;
            margin: -20px 0 0 32px;
          }
        }

        .action-label {
          display: contents;
          position: absolute;
          margin-top: 5px;

          span {
            display: block;
            margin: -19px 0 0 32px;
          }
        }

        svg {
          width: 22px;
          height: 22px;
        }

        &:hover {
          background-color: $first-color-light;
          font-family: $font-family;
        }
      }

      .taxonomy-label-border {
        padding: 5px 10px 10px;
        border-bottom: 1px solid $grey1;
      }

      .taxonomy-category {
        font-family: $font-family3;
        padding: 25px 10px 10px;
      }

      .taxonomy-label.active {
        background-color: $first-color-light;
        font-family: $font-family;
      }
    }
  }

  .harvesting-box-body {
    padding: 20px;
    font-size: $font-size-xs;
    color: $grey3;
    word-wrap: break-word;

    .form-group {
      margin: 0;
    }

    textarea {
      border: 1px solid $grey1;
      margin: -20px 0 0;
      color: $grey2;
      font-family: $font-family;
      font-size: $font-size-xs;
      min-height: 100px;

      &:focus {
        border: 1px solid $grey1;
      }
    }
  }

  .harvesting-box-footer {
    text-align: center;
    padding: 20px 0;
    background-color: $grey1;
    display: flex;
    justify-content: space-around;

    .btn-default {
      margin: 0 22px;
    }

    .button-submit {
      padding: 8px 16px;
    }

    .button-cancel {
      padding: 8px 16px;
    }
  }
}

.extract-in-message {
  display: inline;
}

.harvesting-mode-on {
  .extract-in-message {
    background-color: $harvesting-green-light;

    &.submitted {
      background-color: $harvesting-blue-light;
    }
  }
}

.submitted-harvesting {
  .harvesting-submitted-message {
    color: $harvesting-blue;
    padding-top: 10px;
    font-size: 12px;
    padding-left: 8px;
    margin-bottom: -10px;
  }

  .validated-harvesting {
    color: $harvesting-blue !important;
  }

  .box-icon {
    border: 3px solid $harvesting-blue;
  }

  .active-box {
    border-left-color: $harvesting-blue;
  }
}

@media screen and (max-width: 767px) {
  .is-harvesting-button {
    display: none;
  }

  .harvesting-container {
    display: none;
  }
}
