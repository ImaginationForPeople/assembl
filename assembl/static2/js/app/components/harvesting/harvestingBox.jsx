// @flow
import * as React from 'react';
import ARange from 'annotator_range'; // eslint-disable-line
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Button, OverlayTrigger } from 'react-bootstrap';
import { Translate, I18n } from 'react-redux-i18n';
import classnames from 'classnames';
import moment from 'moment';

import addPostExtractMutation from '../../graphql/mutations/addPostExtract.graphql';
import updateExtractMutation from '../../graphql/mutations/updateExtract.graphql';
import deleteExtractMutation from '../../graphql/mutations/deleteExtract.graphql';
import confirmExtractMutation from '../../graphql/mutations/confirmExtract.graphql';
import withLoadingIndicator from '../../components/common/withLoadingIndicator';
import { getConnectedUserId, getConnectedUserName } from '../../utils/globalFunctions';
import AvatarImage from '../common/avatarImage';
import TaxonomyOverflowMenu from './taxonomyOverflowMenu';
import FormControlWithLabel from '../common/formControlWithLabel';
import { displayAlert, displayModal, closeModal } from '../../utils/utilityManager';
import { editExtractTooltip, deleteExtractTooltip, nuggetExtractTooltip, qualifyExtractTooltip } from '../common/tooltips';
import { NatureIcons, ActionIcons } from '../../utils/extractQualifier';
import { ExtractStates } from '../../constants';

type Props = {
  extract: ?Extract,
  postId: string,
  contentLocale: string,
  lang?: string,
  selection: ?Object,
  setHarvestingBoxDisplay: Function,
  cancelHarvesting: Function,
  addPostExtract: Function,
  updateExtract: Function,
  confirmExtract: Function,
  deleteExtract: Function,
  refetchPost: Function,
  harvestingDate?: string,
  isAuthorAccountDeleted?: boolean
};

type State = {
  disabled: boolean,
  extractIsValidated: boolean,
  isNugget: boolean,
  isEditable: boolean,
  editableExtract: string,
  extractNature: ?string,
  extractAction: ?string,
  showOverflowMenu: boolean,
  overflowMenu: ?HTMLElement,
  overflowMenuTop: number
};

type Taxonomies = {
  nature: ?string,
  action: ?string
};

const ACTIONS = {
  create: 'create', // create a new extract
  edit: 'edit', // edit an extract
  confirm: 'confirm' // confirm a submitted extract
};

class DumbHarvestingBox extends React.Component<Props, State> {
  menu: any;

  actions: any;

  static defaultProps = {
    harvestingDate: null,
    isAuthorAccountDeleted: false
  };

  constructor(props: Props) {
    super(props);
    const { extract, cancelHarvesting } = this.props;
    const isExtract = extract !== null;
    const isNugget = extract ? extract.important : false;
    this.state = {
      disabled: !isExtract,
      extractIsValidated: isExtract,
      isNugget: isNugget,
      isEditable: false,
      editableExtract: extract ? extract.body : '',
      extractNature: extract && extract.extractNature ? extract.extractNature.split('.')[1] : null,
      extractAction: extract && extract.extractAction ? extract.extractAction.split('.')[1] : null,
      showOverflowMenu: false,
      overflowMenu: null,
      overflowMenuTop: 25
    };
    // actions props
    this.actions = {
      [ACTIONS.create]: {
        buttons: [
          { id: 'cancel', title: 'debate.confirmDeletionButtonCancel', className: 'button-cancel', onClick: cancelHarvesting },
          { id: 'validate', title: 'common.attachFileForm.submit', className: 'button-submit', onClick: this.validateHarvesting }
        ]
      },
      [ACTIONS.edit]: {
        buttons: [
          { id: 'cancel', title: 'debate.confirmDeletionButtonCancel', className: 'button-cancel', onClick: this.setEditMode },
          {
            id: 'validate',
            title: 'common.attachFileForm.submit',
            className: 'button-submit',
            onClick: this.updateHarvestingBody
          }
        ]
      },
      [ACTIONS.confirm]: {
        buttons: [
          { id: 'reject', title: 'harvesting.reject', className: 'button-cancel', onClick: this.deleteHarvesting },
          { id: 'confirm', title: 'harvesting.confirm', className: 'button-submit', onClick: this.confirmHarvesting }
        ]
      }
    };
  }

  componentDidMount() {
    window.addEventListener('scroll', this.updateOverflowMenuPosition);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.updateOverflowMenuPosition);
  }

  setEditMode = (): void => {
    const { isEditable } = this.state;
    this.setState({ isEditable: !isEditable });
  };

  editExtract = (value: string): void => {
    this.setState({ editableExtract: value });
  };

  qualifyExtract = (taxonomies: Taxonomies): void => {
    this.setState({ showOverflowMenu: false });
    const { nature, action } = taxonomies;
    const { extract, updateExtract, refetchPost } = this.props;
    const { isNugget } = this.state;
    const variables = {
      extractId: extract ? extract.id : null,
      important: isNugget,
      extractNature: nature,
      extractAction: action
    };
    displayAlert('success', I18n.t('loading.wait'));
    updateExtract({ variables: variables })
      .then(() => {
        this.setState({ extractNature: nature, extractAction: action });
        displayAlert('success', I18n.t('harvesting.harvestingSuccess'));
        refetchPost();
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  updateHarvestingNugget = (): void => {
    const { extract, updateExtract, refetchPost } = this.props;
    const { isNugget, extractNature, extractAction } = this.state;
    const variables = {
      extractId: extract ? extract.id : null,
      important: !isNugget,
      extractNature: extractNature,
      extractAction: extractAction
    };
    displayAlert('success', I18n.t('loading.wait'));
    updateExtract({ variables: variables })
      .then(() => {
        this.setState({
          isNugget: !isNugget
        });
        displayAlert('success', I18n.t('harvesting.harvestingSuccess'));
        refetchPost();
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  updateHarvestingBody = (): void => {
    const { extract, updateExtract, refetchPost } = this.props;
    const { editableExtract, isNugget, extractNature, extractAction } = this.state;
    const variables = {
      extractId: extract ? extract.id : null,
      body: editableExtract,
      important: isNugget,
      extractNature: extractNature,
      extractAction: extractAction
    };
    displayAlert('success', I18n.t('loading.wait'));
    updateExtract({ variables: variables })
      .then(() => {
        this.setState({
          isEditable: false
        });
        displayAlert('success', I18n.t('harvesting.harvestingSuccess'));
        refetchPost();
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  confirmHarvestingDeletion = (): void => {
    const modalTitle = <Translate value="harvesting.deleteExtract" />;
    const body = <Translate value="harvesting.confirmDeleteExtract" />;
    const footer = [
      <Button key="cancel" onClick={closeModal} className="button-cancel button-dark">
        <Translate value="debate.confirmDeletionButtonCancel" />
      </Button>,
      <Button key="delete" onClick={this.deleteHarvesting} className="button-submit button-dark">
        <Translate value="validate" />
      </Button>
    ];
    const includeFooter = true;
    return displayModal(modalTitle, body, includeFooter, footer);
  };

  deleteHarvesting = (): void => {
    const { extract, deleteExtract, refetchPost } = this.props;
    const variables = {
      extractId: extract ? extract.id : null
    };
    closeModal();
    displayAlert('success', I18n.t('loading.wait'));
    deleteExtract({ variables: variables })
      .then(() => {
        displayAlert('success', I18n.t('harvesting.harvestingDeleted'));
        refetchPost();
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  validateHarvesting = (): void => {
    const { postId, selection, contentLocale, lang, addPostExtract, setHarvestingBoxDisplay, refetchPost } = this.props;
    if (!selection) {
      return;
    }
    const selectionText = selection.toString();
    const annotatorRange = ARange.sniff(selection.getRangeAt(0));
    if (!annotatorRange) {
      return;
    }
    const serializedAnnotatorRange = annotatorRange.serialize(document, 'annotation');
    if (!serializedAnnotatorRange) {
      return;
    }
    const variables = {
      contentLocale: contentLocale,
      postId: postId,
      body: selectionText,
      important: false,
      lang: lang,
      xpathStart: serializedAnnotatorRange.start,
      xpathEnd: serializedAnnotatorRange.end,
      offsetStart: serializedAnnotatorRange.startOffset,
      offsetEnd: serializedAnnotatorRange.endOffset
    };
    displayAlert('success', I18n.t('loading.wait'));
    addPostExtract({ variables: variables })
      .then(() => {
        this.setState({
          disabled: false,
          extractIsValidated: true
        });
        setHarvestingBoxDisplay();
        window.getSelection().removeAllRanges();
        refetchPost();
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  confirmHarvesting = (): void => {
    const { extract, confirmExtract, refetchPost } = this.props;
    const variables = {
      extractId: extract ? extract.id : null
    };
    closeModal();
    displayAlert('success', I18n.t('loading.wait'));
    confirmExtract({ variables: variables })
      .then(() => {
        displayAlert('success', I18n.t('harvesting.harvestingConfirmed'));
        refetchPost();
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  showValidatedHarvesting = (nature: ?string, action: ?string) => {
    if (nature && action) {
      return (
        <div className="harvesting-taxonomy-label">
          {`${I18n.t(`search.taxonomy_nature.${nature}`)} + ${I18n.t(`search.taxonomy_action.${action}`)}`}
        </div>
      );
    } else if (nature) {
      return <div className="harvesting-taxonomy-label">{I18n.t(`search.taxonomy_nature.${nature}`)}</div>;
    }
    return action ? <div className="harvesting-taxonomy-label">{I18n.t(`search.taxonomy_action.${action}`)}</div> : null;
  };

  updateOverflowMenu = (node: HTMLElement) => {
    if (node) {
      this.setState({ overflowMenu: node });
    }
  };

  updateOverflowMenuPosition = () => {
    const { overflowMenu } = this.state;
    if (overflowMenu) {
      const height = overflowMenu.clientHeight;
      const bottomScroll = window.pageYOffset + height;
      const windowHeight = document.body && document.body.scrollHeight;
      const isBottomReached = windowHeight && bottomScroll >= windowHeight - window.innerHeight;
      if (isBottomReached) {
        this.setState({ overflowMenuTop: -320 });
      } else {
        this.setState({ overflowMenuTop: 25 });
      }
    }
  };

  renderFooter = () => {
    const { extract } = this.props;
    const { disabled, isEditable } = this.state;
    const extractState = extract && extract.extractState;
    const isSubmitted = extractState === ExtractStates.SUBMITTED;
    const actionId = isEditable ? ACTIONS.edit : (disabled && ACTIONS.create) || (isSubmitted && ACTIONS.confirm);
    if (!actionId) return null;
    const action = this.actions[actionId];
    return (
      <div className="harvesting-box-footer">
        {action.buttons.map(button => (
          <Button key={button.id} className={`${button.className} button-dark`} onClick={button.onClick}>
            {I18n.t(button.title)}
          </Button>
        ))}
      </div>
    );
  };

  render() {
    const { selection, extract, contentLocale, harvestingDate, isAuthorAccountDeleted } = this.props;
    const {
      disabled,
      extractIsValidated,
      isNugget,
      isEditable,
      editableExtract,
      extractNature,
      extractAction,
      showOverflowMenu,
      overflowMenuTop
    } = this.state;
    const isExtract = extract !== null;
    const selectionText = selection ? selection.toString() : '';
    const harvesterUserName =
      extract && extract.creator && extract.creator.displayName ? extract.creator.displayName : getConnectedUserName();
    const extractState = extract && extract.extractState;
    const isSubmitted = extractState === ExtractStates.SUBMITTED;
    const userName = isAuthorAccountDeleted ? I18n.t('deletedUser') : harvesterUserName;
    const harvesterUserId = extract && extract.creator && extract.creator.userId ? extract.creator.userId : getConnectedUserId();
    const menuDisabled = disabled || isSubmitted;
    const hasFooter = disabled || isEditable || isSubmitted;
    return (
      <div className={isSubmitted ? 'submitted-harvesting' : ''}>
        {(extractNature || extractAction) && (
          <div>
            <div className="box-icon">
              {extractNature && <NatureIcons qualifier={extractNature} />}
              {extractAction && !extractNature && <ActionIcons qualifier={extractAction} backgroundColor="#fff" color="#000" />}
            </div>
            {extractNature &&
              extractAction && (
                <div className="box-icon box-icon-2">
                  <ActionIcons qualifier={extractAction} backgroundColor="#fff" color="#000" />
                </div>
              )}
          </div>
        )}
        <div
          className={classnames('theme-box', 'harvesting-box', {
            'active-box': extractIsValidated
          })}
        >
          <div className="harvesting-box-header">
            <div className="harvesting-status">
              {disabled ? (
                <div className="harvesting-in-progress">
                  <div className="harvesting-status-label">
                    <Translate value="harvesting.inProgress" />
                  </div>
                </div>
              ) : (
                <div>
                  {extractNature || extractAction ? (
                    <div className="validated-harvesting">{this.showValidatedHarvesting(extractNature, extractAction)}</div>
                  ) : (
                    !isSubmitted && (
                      <div className="validated-harvesting">
                        <div className="harvesting-status-label">
                          <Translate value="harvesting.validated" />
                        </div>
                      </div>
                    )
                  )}
                </div>
              )}
            </div>
            <div className="button-bar">
              <OverlayTrigger placement="top" overlay={editExtractTooltip}>
                <Button disabled={menuDisabled} onClick={this.setEditMode} className={classnames({ active: isEditable })}>
                  <span className="assembl-icon-edit grey" />
                </Button>
              </OverlayTrigger>
              <OverlayTrigger placement="top" overlay={deleteExtractTooltip}>
                <Button disabled={menuDisabled} onClick={this.confirmHarvestingDeletion}>
                  <span className="assembl-icon-delete grey" />
                </Button>
              </OverlayTrigger>
              <OverlayTrigger placement="top" overlay={nuggetExtractTooltip}>
                <Button
                  disabled={menuDisabled}
                  onClick={this.updateHarvestingNugget}
                  className={classnames({ active: isNugget })}
                >
                  <span className="assembl-icon-pepite grey" />
                </Button>
              </OverlayTrigger>
              <OverlayTrigger placement="top" overlay={qualifyExtractTooltip}>
                <Button
                  disabled={menuDisabled}
                  className="taxonomy-menu-btn"
                  onClick={() => {
                    this.setState({ showOverflowMenu: !showOverflowMenu });
                  }}
                >
                  <span className="assembl-icon-ellipsis-vert grey" />
                </Button>
              </OverlayTrigger>
              {showOverflowMenu && (
                <TaxonomyOverflowMenu
                  innerRef={this.updateOverflowMenu}
                  handleClick={this.qualifyExtract}
                  extractNature={extractNature}
                  extractAction={extractAction}
                  onCloseClick={() => {
                    this.setState({ showOverflowMenu: false });
                  }}
                  top={overflowMenuTop}
                />
              )}
            </div>
            <div className="profile">
              <AvatarImage userId={harvesterUserId} userName={userName} />
              <div className="harvesting-infos">
                <div className="username">{userName}</div>
                {isExtract &&
                  extract &&
                  extract.creationDate && (
                    <div className="harvesting-date" title={extract.creationDate}>
                      {harvestingDate ||
                        moment(extract.creationDate)
                          .locale(contentLocale)
                          .fromNow()}
                    </div>
                  )}
                {!isExtract && (
                  <div className="harvesting-date">
                    <Translate value="harvesting.now" />
                  </div>
                )}
              </div>
            </div>
          </div>
          {isSubmitted && (
            <div className="harvesting-submitted-message">
              <span className="confirm-harvest-button assembl-icon-catch" />
              <Translate value="harvesting.harvestingSubmitted" />
            </div>
          )}
          <div className="harvesting-box-body">
            {isExtract && extract && !isEditable && <div>{extract.body}</div>}
            {isExtract &&
              extract &&
              isEditable && (
                <FormControlWithLabel
                  label=""
                  componentClass="textarea"
                  className="text-area"
                  value={editableExtract}
                  onChange={e => this.editExtract(e.target.value)}
                />
              )}
            {!isExtract && <div>{selectionText}</div>}
          </div>
          {hasFooter && this.renderFooter()}
        </div>
      </div>
    );
  }
}

export { DumbHarvestingBox };

const mapStateToProps = state => ({
  contentLocale: state.i18n.locale
});

export default compose(
  connect(mapStateToProps),
  graphql(addPostExtractMutation, {
    name: 'addPostExtract'
  }),
  graphql(updateExtractMutation, {
    name: 'updateExtract'
  }),
  graphql(deleteExtractMutation, {
    name: 'deleteExtract'
  }),
  graphql(confirmExtractMutation, {
    name: 'confirmExtract'
  }),
  withLoadingIndicator()
)(DumbHarvestingBox);