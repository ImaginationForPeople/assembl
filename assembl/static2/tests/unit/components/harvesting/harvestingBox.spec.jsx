import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbHarvestingBox } from '../../../../js/app/components/harvesting/harvestingBox';
import * as fakeData from './fakeData';

describe('harvestingBox component', () => {
  it('should match harvestingBox snapshot', () => {
    const { extract } = fakeData;
    const props = {
      postId: '12345',
      extract: extract,
      index: 0,
      previousExtractId: '876876876',
      harvestingAnchorPosition: { x: 100, y: 200 },
      contentLocale: 'fr',
      harvestingDate: 'il y a 5 jours'
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbHarvestingBox {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
  it('should match submitted harvestingBox snapshot', () => {
    const { submittedExtract } = fakeData;
    const props = {
      postId: '12345',
      extract: submittedExtract,
      index: 0,
      previousExtractId: '876876876',
      harvestingAnchorPosition: { x: 100, y: 200 },
      contentLocale: 'fr',
      harvestingDate: 'il y a 5 jours'
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbHarvestingBox {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});