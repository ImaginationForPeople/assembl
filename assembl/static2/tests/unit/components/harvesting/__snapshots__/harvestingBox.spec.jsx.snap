// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`harvestingBox component should match harvestingBox snapshot 1`] = `
<div
  className=""
>
  <div
    className="theme-box harvesting-box active-box"
  >
    <div
      className="harvesting-box-header"
    >
      <div
        className="harvesting-status"
      >
        <div>
          <div
            className="validated-harvesting"
          >
            <div
              className="harvesting-status-label"
            >
              <Translate
                value="harvesting.validated"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="button-bar"
      >
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="editExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.editExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className=""
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-edit grey"
            />
          </Button>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="deleteExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.deleteExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-delete grey"
            />
          </Button>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="nuggetExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.nuggetExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className=""
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-pepite grey"
            />
          </Button>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="qualifyExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.qualifyExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="taxonomy-menu-btn"
            disabled={false}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-ellipsis-vert grey"
            />
          </Button>
        </OverlayTrigger>
      </div>
      <div
        className="profile"
      >
        <AvatarImage
          userId={31}
          userName="John Doe"
        />
        <div
          className="harvesting-infos"
        >
          <div
            className="username"
          >
            John Doe
          </div>
          <div
            className="harvesting-date"
            title="2018-03-29T16:28:27.324276+00:00"
          >
            il y a 5 jours
          </div>
        </div>
      </div>
    </div>
    <div
      className="harvesting-box-body"
    >
      <div>
        Hello world!
      </div>
    </div>
  </div>
</div>
`;

exports[`harvestingBox component should match submitted harvestingBox snapshot 1`] = `
<div
  className="submitted-harvesting"
>
  <div
    className="theme-box harvesting-box active-box"
  >
    <div
      className="harvesting-box-header"
    >
      <div
        className="harvesting-status"
      >
        <div />
      </div>
      <div
        className="button-bar"
      >
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="editExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.editExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className=""
            disabled={true}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-edit grey"
            />
          </Button>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="deleteExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.deleteExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={true}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-delete grey"
            />
          </Button>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="nuggetExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.nuggetExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className=""
            disabled={true}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-pepite grey"
            />
          </Button>
        </OverlayTrigger>
        <OverlayTrigger
          defaultOverlayShown={false}
          overlay={
            <Tooltip
              bsClass="tooltip"
              id="qualifyExtractTooltip"
              placement="right"
            >
              <Translate
                value="harvesting.qualifyExtract"
              />
            </Tooltip>
          }
          placement="top"
          trigger={
            Array [
              "hover",
              "focus",
            ]
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="taxonomy-menu-btn"
            disabled={true}
            onClick={[Function]}
          >
            <span
              className="assembl-icon-ellipsis-vert grey"
            />
          </Button>
        </OverlayTrigger>
      </div>
      <div
        className="profile"
      >
        <AvatarImage
          userId={31}
          userName="John Doe"
        />
        <div
          className="harvesting-infos"
        >
          <div
            className="username"
          >
            John Doe
          </div>
          <div
            className="harvesting-date"
            title="2018-03-29T16:28:27.324276+00:00"
          >
            il y a 5 jours
          </div>
        </div>
      </div>
    </div>
    <div
      className="harvesting-submitted-message"
    >
      <span
        className="confirm-harvest-button assembl-icon-catch"
      />
      <Translate
        value="harvesting.harvestingSubmitted"
      />
    </div>
    <div
      className="harvesting-box-body"
    >
      <div>
        Hello world!
      </div>
    </div>
    <div
      className="harvesting-box-footer"
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="button-cancel button-dark"
        disabled={false}
        onClick={[Function]}
      >
        Reject
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="button-submit button-dark"
        disabled={false}
        onClick={[Function]}
      >
        Confirm
      </Button>
    </div>
  </div>
</div>
`;
